timemachine/md/smc.py

```diff
@@ -120,6 +123,153 @@ def accumulate_results(samples, indices, log_weights, incremental_log_weights):
+
+    References
+    ----------
+    * [Zhou, Johansen, Aston, 2016]
+        Towards Automatic Model Comparison: An Adaptive Sequential Monte Carlo Approach
+        https://arxiv.org/pdf/1303.3123 (Algorithm #4)
+    """
+    n = len(samples)
+
+    # check cess_target
+    assert cess_target > 1, f"cess_target is too small: {cess_target} <= 1"
+    assert cess_target < n, f"cess_target is too large: {cess_target} >= {n}"
+
+    log_weights = np.zeros(n)
+    norm_log_weights = log_weights - logsumexp(log_weights)
+
+    # store
+    sample_traj = [samples]
+    ancestry_traj = [np.arange(n)]
+    log_weights_traj = [np.array(log_weights)]
+    incremental_log_weights_traj = []  # note: redundant but convenient
+    lambdas_traj = [0.0]
+
+    def accumulate_results(samples, indices, log_weights, incremental_log_weights, lam_target):
+        if store_intermediate_traj:
+            sample_traj.append(samples)
+        else:
+            # only store one intermediate set of samples to reduce memory usage
+            sample_traj[0] = samples
+        ancestry_traj.append(indices)
+        log_weights_traj.append(np.array(log_weights))
+        incremental_log_weights_traj.append(np.array(incremental_log_weights))
+        lambdas_traj.append(lam_target)
+
+    lam_initial = 0.0
+    lam_target = 1.0  # adapted
+
+    # Main ASMC loop
+    while True:
+
+        # binary search for lambda that gives cess ~= cess_target
+        cur_log_prob = log_prob(sample_traj[-1], lam_initial, True)
+
+        # Used to pass incremental_log_weights out of the closure
+        incremental_log_weights_closure = [None]
+
+        def f_opt(lam: float) -> float:
+            incremental_log_weights_closure[0] = log_prob(sample_traj[-1], lam, False) - cur_log_prob
+            cess = conditional_effective_sample_size(norm_log_weights, incremental_log_weights_closure[0])
+            return cess - cess_target
+
+        try:
+            lam_target = root_scalar(f_opt, bracket=(lam_initial, lam_target), method="bisect", xtol=epsilon).root
+        except ValueError:
+            # no root, just run at the final lambda
+            pass
+
+        incremental_log_weights = incremental_log_weights_closure[0]
+        assert incremental_log_weights is not None
+
+        # Stop when lam_target == 1.0
+        #   See
+        #   * discussion at https://github.com/proteneer/timemachine/pull/718#discussion_r854276326
+        #   * helper function get_endstate_samples_from_smc_result
+        if lam_target == 1.0:
+            break
+
+        # resample
+        indices, log_weights = resample(log_weights + incremental_log_weights)
+        norm_log_weights = log_weights - logsumexp(log_weights)
+        resampled = [sample_traj[-1][i] for i in indices]
+
+        # propagate
+        samples = propagate(resampled, lam_target)
+
+        # log
+        accumulate_results(samples, indices, log_weights, incremental_log_weights, lam_target)
```

Comment:
nit: possible future logging enhancement: forecast # steps remaining

(crudely, could guess this based on the size of the most recent step https://gist.github.com/maxentile/be328e929abf4a92bee7d26967277f54#file-cess_smc_openmm-py-L132-L133 or similar)

Reply:
Perhaps for the future